lib/connection.ts

```diff
@@ -0,0 +1,355 @@
+      rheaSession.begin();
+    });
+  }
+
+  /**
+   * Creates an amqp sender link. It either uses the provided session or creates a new one.
+   * @param {SenderOptionsWithSession} options Optional parameters to create a sender link.
+   * @return {Promise<Sender>} Promise<Sender>.
+   */
+  async createSender(options?: SenderOptionsWithSession): Promise<Sender> {
+    if (options && options.session && options.session.createSender) {
+      return await options.session.createSender(options);
+    }
+    const session = await this.createSession();
+    return await session.createSender(options);
+  }
+
+  /**
+   * Creates an amqp receiver link. It either uses the provided session or creates a new one.
+   * @param {ReceiverOptionsWithSession} options Optional parameters to create a receiver link.
+   * @return {Promise<Receiver>} Promise<Receiver>.
+   */
+  async createReceiver(options?: ReceiverOptionsWithSession): Promise<Receiver> {
+    if (options && options.session && options.session.createReceiver) {
+      return await options.session.createReceiver(options);
+    }
+    const session = await this.createSession();
+    return await session.createReceiver(options);
+  }
+
+  /**
+   * Creates an amqp sender-receiver link. It either uses the provided session or creates a new one.
+   * This method creates a sender-receiver link on the same session. It is useful for management
+   * style operations where one may want to send a request and await for response.
+   * @param {SenderOptions} senderOptions Parameters to create a sender.
+   * @param {ReceiverOptions} receiverOptions Parameters to create a receiver.
+   * @param {Session} [session] The optional session on which the sender and receiver links will be
+   * created.
+   * @return {Promise<ReqResLink>} Promise<ReqResLink>
+   */
+  async createRequestResponseLink(senderOptions: SenderOptions, receiverOptions: ReceiverOptions, providedSession?: Session): Promise<ReqResLink> {
+    if (!senderOptions) {
+      throw new Error(`Please provide sender options.`);
+    }
+    if (!receiverOptions) {
+      throw new Error(`Please provide receiver options.`);
+    }
+    const session = providedSession || await this.createSession();
+    const sender = await session.createSender(senderOptions);
+    const receiver = await session.createReceiver(receiverOptions);
+    log.connection("[%s] Successfully created the sender and receiver links on the same session.", this.id);
+    return {
+      session: session,
+      sender: sender,
+      receiver: receiver
+    };
+  }
+
+  /**
+   * Registers the given handler for the specified event.
+   * @param {rhea.ConnectionEvents} event The event for which the handler needs to be registered.
+   * @param {rhea.OnAmqpEvent} handler The handler function that needs to be executed when the event
+   * occurs.
+   * @returns {void} void.
+   */
+  registerHandler(event: ConnectionEvents, handler: rhea.OnAmqpEvent): void {
```

Comment:
why not expose this as a standard event emitter instead of using specially-named functions?

Reply:
i think it makes sense to because it's functioning like an event emitter even if you're not calling it one. that way the APIs are consistent with what people do elsewhere